velox/dwio/dwrf/writer/ColumnWriter.cpp

```diff
@@ -697,6 +697,89 @@ class TimestampColumnWriter : public BaseColumnWriter {
   std::unique_ptr<IntEncoder<false>> nanos_;
 };
 
+class DecimalColumnWriter : public BaseColumnWriter {
+ public:
+  DecimalColumnWriter(
+      WriterContext& context,
+      const TypeWithId& type,
+      uint32_t sequence,
+      std::function<void(IndexBuilder&)> onRecordPosition)
+      : BaseColumnWriter{context, type, sequence, onRecordPosition},
+        type_{type.type()},
+        unscaledValues_{createDirectEncoder</* isSigned = */ true>(
+            newStream(StreamKind::StreamKind_DATA),
+            getConfig(Config::USE_VINTS),
+            type_->isShortDecimal() ? LONG_BYTE_SIZE : 2 * LONG_BYTE_SIZE)},
+        scales_{createRleEncoder</* isSigned = */ true>(
+            RleVersion_1,
+            // DWRF's NANO_DATA has the same enum value as ORC's SECONDARY.
+            newStream(StreamKind::StreamKind_NANO_DATA),
+            context.getConfig(Config::USE_VINTS),
+            LONG_BYTE_SIZE)} {
+    reset();
+  }
+
+  uint64_t write(const VectorPtr& slice, const common::Ranges& ranges)
+      override {
+    VELOX_CHECK(
+        slice->type()->equivalent(*type_),
+        "Unexpected vector type: {}.",
+        slice->type()->toString());
+    auto localDecoded = decode(slice, ranges);
+    auto& decodedVector = localDecoded.get();
+    writeNulls(decodedVector, ranges);
+
+    size_t count = 0;
+    auto [_, scale] = getDecimalPrecisionScale(*type_);
+
+    for (auto& pos : ranges) {
+      if (!decodedVector.mayHaveNulls() || !decodedVector.isNullAt(pos)) {
+        if (type_->isShortDecimal()) {
+          auto val = decodedVector.valueAt<int64_t>(pos);
+          unscaledValues_->writeValue(val);
+          scales_->writeValue(scale);
```

Comment:
I wonder if we need to truncate the zeros before writing, since I notice that the decimal reader could recover a decimal value to the target scale with the recorded 'scales'. For example, 123.000 of scale 3 could be written as 123'000 with scale 3 or 123 with scale 0. To truncate the tailing zeros might benefits for the compression ratio and helps reduce file size, but it requires extra conversion and could increase write time. @Yuhta Do you have any suggestion? Thanks!

https://github.com/facebookincubator/velox/blob/d1bf9da3d4d36685569578eacea771088c1c9330/velox/type/DecimalUtil.h#L108-L114

Reply:
What does the other file format do for this? We could keep this simple to optimize processing time.

Reply:
From compute engine standpoint we would prefer process time over small storage size.  Both alternatives generate valid files though, I think we can use whatever is simpler to implement for now and come back later if we see there is need to change.

Reply:
I notice Parquet does not contain a scales vector. Thanks for your suggestion. I keep it this way.